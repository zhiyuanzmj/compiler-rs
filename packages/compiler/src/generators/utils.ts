import {
  advancePositionWithMutation,
  locStub,
  NewlineType,
  type CodegenSourceMapGenerator,
  type Position,
  type SourceLocation,
} from '@vue/compiler-dom'
import { isArray, isString } from '@vue/shared'
import { SourceMapGenerator } from 'source-map-js'
import type { CodegenContext } from '../generate'

export { genExpression } from './expression'

export const NEWLINE: unique symbol = Symbol(`newline`)
export const INDENT_START: unique symbol = Symbol(`indent start`)
export const INDENT_END: unique symbol = Symbol(`indent end`)

type FalsyValue = false | null | undefined
export type CodeFragment =
  | typeof NEWLINE
  | typeof INDENT_START
  | typeof INDENT_END
  | string
  | [code: string, newlineIndex?: number, loc?: SourceLocation, name?: string]
  | FalsyValue
export type CodeFragments = Exclude<CodeFragment, any[]> | CodeFragment[]

export function buildCodeFragment(
  ...frag: CodeFragment[]
): [
  CodeFragment[],
  (...items: CodeFragment[]) => number,
  (...items: CodeFragment[]) => number,
] {
  const push = frag.push.bind(frag)
  const unshift = frag.unshift.bind(frag)
  return [frag, push, unshift]
}

export type CodeFragmentDelimiters = [
  left: CodeFragments,
  right: CodeFragments,
  delimiter: CodeFragments,
  placeholder?: CodeFragments,
]

export function genMulti(
  [left, right, seg, placeholder]: CodeFragmentDelimiters,
  ...frags: CodeFragments[]
): CodeFragment[] {
  if (placeholder) {
    while (frags.length > 0 && !frags.at(-1)) {
      frags.pop()
    }
    frags = frags.map((frag) => frag || placeholder)
  } else {
    frags = frags.filter(Boolean)
  }

  const frag: CodeFragment[] = []
  push(left)
  for (const [i, fn] of (
    frags as Array<Exclude<CodeFragments, FalsyValue>>
  ).entries()) {
    push(fn)
    if (i < frags.length - 1) push(seg)
  }
  push(right)
  return frag

  function push(fn: CodeFragments) {
    if (!isArray(fn)) fn = [fn]
    frag.push(...fn)
  }
}
export const DELIMITERS_ARRAY: CodeFragmentDelimiters = ['[', ']', ', ']
export const DELIMITERS_ARRAY_NEWLINE: CodeFragmentDelimiters = [
  ['[', INDENT_START, NEWLINE],
  [INDENT_END, NEWLINE, ']'],
  [', ', NEWLINE],
]
export const DELIMITERS_OBJECT: CodeFragmentDelimiters = ['{ ', ' }', ', ']
export const DELIMITERS_OBJECT_NEWLINE: CodeFragmentDelimiters = [
  ['{', INDENT_START, NEWLINE],
  [INDENT_END, NEWLINE, '}'],
  [', ', NEWLINE],
]

export function genCall(
  name: string | [name: string, placeholder?: CodeFragments],
  ...frags: CodeFragments[]
): CodeFragment[] {
  const hasPlaceholder = isArray(name)
  const fnName = hasPlaceholder ? name[0] : name
  const placeholder = hasPlaceholder ? name[1] : 'null'
  return [fnName, ...genMulti(['(', ')', ', ', placeholder], ...frags)]
}

export function codeFragmentToString(
  code: CodeFragment[],
  context: CodegenContext,
): [code: string, map: CodegenSourceMapGenerator | undefined] {
  const {
    options: { filename, sourceMap },
  } = context

  let map: CodegenSourceMapGenerator | undefined
  if (sourceMap) {
    // lazy require source-map implementation, only in non-browser builds
    map = new SourceMapGenerator() as unknown as CodegenSourceMapGenerator
    map.setSourceContent(filename, context.ir.source)
    map._sources.add(filename)
  }

  let codegen = ''
  const pos = { line: 1, column: 1, offset: 0 }
  let indentLevel = 0

  for (let frag of code) {
    if (!frag) continue

    if (frag === NEWLINE) {
      frag = [`\n${`  `.repeat(indentLevel)}`, NewlineType.Start]
    } else if (frag === INDENT_START) {
      indentLevel++
      continue
    } else if (frag === INDENT_END) {
      indentLevel--
      continue
    }

    if (isString(frag)) frag = [frag]

    let [code, newlineIndex = NewlineType.None, loc, name] = frag
    codegen += code

    if (map) {
      if (loc) addMapping(loc.start, name)
      if (newlineIndex === NewlineType.Unknown) {
        // multiple newlines, full iteration
        advancePositionWithMutation(pos, code)
      } else {
        // fast paths
        pos.offset += code.length
        if (newlineIndex === NewlineType.None) {
          pos.column += code.length
        } else {
          // single newline at known index
          if (newlineIndex === NewlineType.End) {
            newlineIndex = code.length - 1
          }
          pos.line++
          pos.column = code.length - newlineIndex
        }
      }
      if (loc && loc !== locStub) {
        addMapping(loc.end)
      }
    }
  }

  return [codegen, map]

  function addMapping(loc: Position, name: string | null = null) {
    // we use the private property to directly add the mapping
    // because the addMapping() implementation in source-map-js has a bunch of
    // unnecessary arg and validation checks that are pure overhead in our case.
    const { _names, _mappings } = map!
    if (name !== null && !_names.has(name)) _names.add(name)
    _mappings.add({
      originalLine: loc.line,
      originalColumn: loc.column - 1, // source-map column is 0 based
      generatedLine: pos.line,
      generatedColumn: pos.column - 1,
      source: filename,
      name,
    })
  }
}
