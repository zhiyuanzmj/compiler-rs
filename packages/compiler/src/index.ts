export { compile, type CompilerOptions, type TransformPreset } from './compile'
export * from './transform'
export {
  CodegenContext,
  generate,
  type CodegenOptions,
  type VaporCodegenResult,
} from './generate'

export * from './ir'

export { transformText } from './transforms/transformText'
export { transformElement } from './transforms/transformElement'
export { transformChildren } from './transforms/transformChildren'
export { transformTemplateRef } from './transforms/transformTemplateRef'
export { transformVBind } from './transforms/vBind'
export { transformVOn } from './transforms/vOn'
export { transformVSlot } from './transforms/vSlot'
export { transformVSlots } from './transforms/vSlots'
export { transformVModel } from './transforms/vModel'
export { transformVShow } from './transforms/vShow'
export { transformVHtml } from './transforms/vHtml'
export { transformVFor } from './transforms/vFor'
export { transformVIf } from './transforms/vIf'
export { transformVOnce } from './transforms/vOnce'
export { transformVText } from './transforms/vText'
