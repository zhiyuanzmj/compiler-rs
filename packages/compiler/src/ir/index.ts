export {
  CreateNodesIRNode,
  _DynamicFlag as DynamicFlag,
  IRNode,
  IRNodeTypes,
  isBlockOperation,
  type BaseIRNode,
  type BlockIRNode,
  type CreateComponentIRNode,
  type DeclareOldRefIRNode,
  type DirectiveIRNode,
  type DirectiveNode,
  type ForIRNode,
  type GetTextChildIRNode,
  type IfIRNode,
  type InsertionStateTypes,
  type InsertNodeIRNode,
  type IRDynamicInfo,
  type IREffect,
  type IRFor,
  type OperationNode,
  type PrependNodeIRNode,
  type RootIRNode,
  type SetDynamicEventsIRNode,
  type SetDynamicPropsIRNode,
  type SetEventIRNode,
  type SetHtmlIRNode,
  type SetNodesIRNode,
  type SetPropIRNode,
  type SetTemplateRefIRNode,
  type SetTextIRNode,
  type SimpleExpressionNode,
  type SourceLocation,
} from '@vue-jsx-vapor/compiler-rs'

export * from './component'
