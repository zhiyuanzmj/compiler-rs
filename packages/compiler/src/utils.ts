export * from './utils/error'
export * from './utils/expression'
export * from './utils/ast'
export * from './utils/directive'
export * from './utils/text'
export * from './utils/transform'
export * from './utils/generate'
export * from './utils/dom'
