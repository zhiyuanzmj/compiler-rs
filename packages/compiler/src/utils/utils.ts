export {
  getExpression,
  getTextLikeValue,
  TS_NODE_TYPES,
  unwrapTSNode,
} from '@vue-jsx-vapor/compiler-rs'
