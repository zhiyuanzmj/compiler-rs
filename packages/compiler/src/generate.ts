import { extend, remove } from '@vue/shared'
import { genBlockContent } from './generators/block'
import { genTemplates } from './generators/template'
import { setTemplateRefIdent } from './generators/templateRef'
import {
  buildCodeFragment,
  codeFragmentToString,
  INDENT_END,
  INDENT_START,
  NEWLINE,
} from './utils/generate'
import type { BlockIRNode, RootIRNode } from './ir'
import type { SimpleExpressionNode } from './utils'
import type { RawSourceMap } from 'source-map-js'

export type CodegenOptions = {
  /**
   * Generate source map?
   * @default false
   */
  sourceMap?: boolean
  /**
   * Filename for source map generation.
   * Also used for self-recursive reference in templates
   * @default 'index.jsx'
   */
  filename?: string
  templates?: string[]
}

export class CodegenContext {
  options: Required<CodegenOptions>

  helpers: Set<string> = new Set<string>([])

  helper = (name: string) => {
    this.helpers.add(name)
    return `_${name}`
  }

  delegates: Set<string> = new Set<string>()

  identifiers: Record<string, (string | SimpleExpressionNode)[]> =
    Object.create(null)

  block: BlockIRNode
  withId<T>(
    fn: () => T,
    map: Record<string, string | SimpleExpressionNode | null>,
  ): T {
    const { identifiers } = this
    const ids = Object.keys(map)

    for (const id of ids) {
      identifiers[id] ||= []
      identifiers[id].unshift(map[id] || id)
    }

    const ret = fn()
    ids.forEach((id) => remove(identifiers[id], map[id] || id))

    return ret
  }

  enterBlock(block: BlockIRNode) {
    const parent = this.block
    this.block = block
    return (): BlockIRNode => (this.block = parent)
  }

  scopeLevel: number = 0
  enterScope(): [level: number, exit: () => number] {
    return [this.scopeLevel++, () => this.scopeLevel--] as const
  }

  constructor(
    public ir: RootIRNode,
    options: CodegenOptions,
  ) {
    const defaultOptions: Required<CodegenOptions> = {
      sourceMap: false,
      filename: `index.jsx`,
      templates: [],
    }
    this.options = extend(defaultOptions, options)
    this.block = ir.block
  }
}

export interface VaporCodegenResult {
  ast: RootIRNode
  helpers: Set<string>
  templates: string[]
  delegates: Set<string>
  code: string
  map?: RawSourceMap
}

// IR -> JS codegen
export function generate(
  ir: RootIRNode,
  options: CodegenOptions = {},
): VaporCodegenResult {
  const [frag, push] = buildCodeFragment()
  const context = new CodegenContext(ir, options)
  const { helpers } = context

  push(INDENT_START)
  if (ir.hasTemplateRef) {
    push(
      NEWLINE,
      `const ${setTemplateRefIdent} = ${context.helper('createTemplateRefSetter')}()`,
    )
  }
  push(...genBlockContent(ir.block, context, true))
  push(INDENT_END, NEWLINE)

  if (context.delegates.size) {
    context.helper('delegateEvents')
  }
  const templates = genTemplates(ir.templates, ir.rootTemplateIndex, context)

  const [code, map] = codeFragmentToString(frag, context)

  return {
    code,
    ast: ir,
    map: map && map.toJSON(),
    helpers,
    templates,
    delegates: context.delegates,
  }
}
