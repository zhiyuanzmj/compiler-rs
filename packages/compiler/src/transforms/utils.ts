import {
  jsxClosingFragment,
  jsxExpressionContainer,
  jsxFragment,
  jsxOpeningFragment,
  type Expression,
  type JSXElement,
  type JSXFragment,
  type Node,
} from '@babel/types'
import { createSimpleExpression } from '@vue/compiler-dom'
import {
  DynamicFlag,
  IRNodeTypes,
  type BlockIRNode,
  type IRDynamicInfo,
  type RootNode,
} from '../ir/index'
import { isTemplate } from '../utils'
import type { TransformContext } from '../transform'

export function newDynamic(): IRDynamicInfo {
  return {
    flags: DynamicFlag.REFERENCED,
    children: [],
  }
}

export function newBlock(node: BlockIRNode['node']): BlockIRNode {
  return {
    type: 1 satisfies IRNodeTypes.BLOCK,
    node,
    dynamic: newDynamic(),
    effect: [],
    operation: [],
    returns: [],
    tempId: 0,
  }
}

export function createBranch(
  node: Parameters<typeof wrapFragment>[0],
  context: TransformContext,
  isVFor?: boolean,
): [BlockIRNode, () => void] {
  context.node = node = wrapFragment(node)
  const branch: BlockIRNode = newBlock(node)
  const exitBlock = context.enterBlock(branch, isVFor)
  context.reference()
  return [branch, exitBlock]
}

export function wrapFragment(node: JSXElement | JSXFragment | Expression) {
  if (node.type === 'JSXFragment' || isTemplate(node)) {
    return node
  }

  return jsxFragment(jsxOpeningFragment(), jsxClosingFragment(), [
    node.type === 'JSXElement' ? node : jsxExpressionContainer(node),
  ])
}

export const EMPTY_EXPRESSION = createSimpleExpression('', true)

export const isFragmentNode = (
  node: Node | RootNode,
): node is JSXElement | JSXFragment | RootNode =>
  node.type === IRNodeTypes.ROOT ||
  node.type === 'JSXFragment' ||
  (node.type === 'JSXElement' && !!isTemplate(node))
