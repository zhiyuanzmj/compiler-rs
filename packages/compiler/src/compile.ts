import { extend, isString } from '@vue/shared'
import { generate, type VaporCodegenResult } from './generate'
import { IRNodeTypes } from './ir'
import { transform, type TransformOptions } from './transform'
import { parseExpression } from './utils'
import type { JSXElement, JSXFragment } from 'oxc-parser'

// code/AST -> IR (transform) -> JS (generate)
export function compile(
  source: JSXElement | JSXFragment | string,
  options: CompilerOptions = {},
): VaporCodegenResult {
  const resolvedOptions = extend({}, options, {
    filename: 'index.jsx',
  })
  if (!resolvedOptions.source && isString(source)) {
    resolvedOptions.source = source
  }
  const root = isString(source)
    ? parseExpression(resolvedOptions.filename, source)
    : source
  const children =
    root.type === 'JSXFragment'
      ? root.children
      : root.type === 'JSXElement'
        ? [root]
        : []
  const ast = {
    type: IRNodeTypes.ROOT,
    children,
  } as unknown as Node

  const ir = transform(ast, extend({}, resolvedOptions))

  return generate(ir, resolvedOptions)
}

export type CompilerOptions = TransformOptions
