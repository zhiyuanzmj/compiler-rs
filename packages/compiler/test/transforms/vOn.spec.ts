import { describe, expect, test, vi } from 'vitest'
import {
  IRNodeTypes,
  transformChildren,
  transformElement,
  transformText,
  transformVBind,
  transformVOn,
} from '../../src'
import { ErrorCodes } from '../../src/utils'
import { makeCompile } from './_utils'

const compileWithVOn = makeCompile({
  nodeTransforms: [transformElement, transformText, transformChildren],
  directiveTransforms: {
    on: transformVOn,
    bind: transformVBind,
  },
})

describe('v-on', () => {
  test('simple expression', () => {
    const { code, ir, helpers } = compileWithVOn(
      `<div onClick={handleClick}></div>`,
    )

    expect(code).toMatchInlineSnapshot(`
      "
        const n0 = t0()
        n0.$evtclick = e => handleClick(e)
        return n0
      "
    `)
    expect(helpers).not.contains('delegate') // optimized as direct attachment
    expect(ir.block.effect).toEqual([])
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.SET_EVENT,
        element: 0,
        key: {
          content: 'click',
          isStatic: true,
        },
        value: {
          content: 'handleClick',
          isStatic: false,
        },
        modifiers: { keys: [], nonKeys: [], options: [] },
        keyOverride: undefined,
        delegate: true,
      },
    ])
  })

  test('event modifier', () => {
    const { code } = compileWithVOn(
      `<>
        <a onClick_stop={handleEvent}></a>
        <form onSubmit_prevent={handleEvent}></form>
        <a onClick_stop_prevent={handleEvent}></a>
        <div onClick_self={handleEvent}></div>
        <div onClick_capture={handleEvent}></div>
        <a onClick_once={handleEvent}></a>
        <div onScroll_passive={handleEvent}></div>
        <input onClick_right={handleEvent} />
        <input onClick_left={handleEvent} />
        <input onClick_middle={handleEvent} />
        <input onClick_enter_right={handleEvent} />
        <input onKeyup_enter={handleEvent} />
        <input onKeyup_tab={handleEvent} />
        <input onKeyup_delete={handleEvent} />
        <input onKeyup_esc={handleEvent} />
        <input onKeyup_space={handleEvent} />
        <input onKeyup_up={handleEvent} />
        <input onKeyup_down={handleEvent} />
        <input onKeyup_left={handleEvent} />
        <input onKeyup_middle={submit} />
        <input onKeyup_middle_self={submit} />
        <input onKeyup_self_enter={handleEvent} />
      </>`,
    )
    expect(code).matchSnapshot()
  })

  test('should error if no expression AND no modifier', () => {
    const onError = vi.fn()
    compileWithVOn(`<div onClick />`, { onError })
    expect(onError.mock.calls[0][0]).toMatchObject({
      code: ErrorCodes.X_V_ON_NO_EXPRESSION,
      loc: {
        start: {
          line: 1,
          column: 5,
        },
        end: {
          line: 1,
          column: 12,
        },
      },
    })
  })

  test('should NOT error if no expression but has modifier', () => {
    const onError = vi.fn()
    compileWithVOn(`<div onClick_prevent />`, { onError })
    expect(onError).not.toHaveBeenCalled()
  })

  test('should support multiple modifiers and event options', () => {
    const { code, ir, helpers } = compileWithVOn(
      `<div onClick_stop_prevent_capture_once={test}/>`,
    )

    expect(code).toMatchSnapshot()
    expect(helpers).contains('on')
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.SET_EVENT,
        value: {
          content: 'test',
          isStatic: false,
        },
        modifiers: {
          keys: [],
          nonKeys: ['stop', 'prevent'],
          options: ['capture', 'once'],
        },
        keyOverride: undefined,
        delegate: false,
      },
    ])
    expect(code).contains(
      `_on(n0, "click", _withModifiers(e => test(e), ["stop","prevent"]), {
    capture: true, 
    once: true
  })`,
    )
  })

  test('should support multiple events and modifiers options', () => {
    const { code, ir } = compileWithVOn(
      `<div onClick_stop={test} onKeyup_enter={test} />`,
    )

    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.SET_EVENT,
        key: {
          content: 'click',
          isStatic: true,
        },
        value: {
          content: 'test',
          isStatic: false,
        },
        modifiers: {
          keys: [],
          nonKeys: ['stop'],
          options: [],
        },
      },
      {
        type: IRNodeTypes.SET_EVENT,
        key: {
          content: 'keyup',
          isStatic: true,
        },
        value: {
          content: 'test',
          isStatic: false,
        },
        modifiers: {
          keys: ['enter'],
          nonKeys: [],
          options: [],
        },
      },
    ])

    expect(code).toMatchInlineSnapshot(`
      "
        const n0 = t0()
        n0.$evtclick = _withModifiers(e => test(e), ["stop"])
        n0.$evtkeyup = _withKeys(e => test(e), ["enter"])
        return n0
      "
    `)
    expect(code).contains(
      `n0.$evtclick = _withModifiers(e => test(e), ["stop"])`,
    )
    expect(code).contains(`n0.$evtkeyup = _withKeys(e => test(e), ["enter"])`)
  })

  test('should wrap keys guard for keyboard events or dynamic events', () => {
    const { code, ir } = compileWithVOn(
      `<div onKeydown_stop_capture_ctrl_a={test}/>`,
    )

    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.SET_EVENT,
        element: 0,
        key: {
          content: 'keydown',
          isStatic: true,
        },
        value: {
          content: 'test',
          isStatic: false,
        },
        modifiers: {
          keys: ['a'],
          nonKeys: ['stop', 'ctrl'],
          options: ['capture'],
        },
      },
    ])

    expect(code).matchSnapshot()
  })

  test('should not wrap keys guard if no key modifier is present', () => {
    const { code, ir } = compileWithVOn(`<div onKeyup_exact={test}/>`)
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.SET_EVENT,
        modifiers: { nonKeys: ['exact'] },
      },
    ])

    expect(code).matchSnapshot()
  })

  test('should wrap keys guard for static key event w/ left/right modifiers', () => {
    const { code, ir } = compileWithVOn(`<div onKeyup_left={test}/>`)

    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.SET_EVENT,
        modifiers: {
          keys: ['left'],
          nonKeys: [],
          options: [],
        },
      },
    ])

    expect(code).matchSnapshot()
  })

  test('should transform click.right', () => {
    const { code, ir, delegates } = compileWithVOn(
      `<div onClick_right={test}/>`,
    )
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.SET_EVENT,
        key: {
          content: 'contextmenu',
          isStatic: true,
        },
        modifiers: { nonKeys: ['right'] },
        keyOverride: undefined,
      },
    ])

    expect(code).toMatchSnapshot()
    expect(delegates).includes('contextmenu')
  })

  test('should transform click.middle', () => {
    const { code, ir, delegates } = compileWithVOn(
      `<div onClick_middle={test}/>`,
    )
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.SET_EVENT,
        key: {
          content: 'mouseup',
          isStatic: true,
        },
        modifiers: { nonKeys: ['middle'] },
        keyOverride: undefined,
      },
    ])

    expect(code).matchSnapshot()
    expect(delegates).includes('mouseup')
  })

  test('should delegate event', () => {
    const { code, ir, helpers, delegates } = compileWithVOn(
      `<div onClick={test}/>`,
    )

    expect(code).matchSnapshot()
    expect(delegates).contains('click')
    expect(helpers).contains('delegateEvents')
    expect(ir.block.operation).toMatchObject([
      {
        type: IRNodeTypes.SET_EVENT,
        delegate: true,
      },
    ])
  })

  test('should use delegate helper when have multiple events of same name', () => {
    const { code, helpers } = compileWithVOn(
      `<div onClick={test} onClick_stop={test} />`,
    )
    expect(helpers).contains('delegate')
    expect(code).toMatchSnapshot()
    expect(code).contains('_delegate(n0, "click", e => test(e))')
    expect(code).contains(
      '_delegate(n0, "click", _withModifiers(e => test(e), ["stop"]))',
    )
  })

  test('namespace event with Component', () => {
    const { code } = compileWithVOn(`<Comp onUpdate:modelValue={() => {}} />`)
    expect(code).toMatchSnapshot()
    expect(code).contains(
      '_createComponent(Comp, { "onUpdate:modelValue": () => () => {} }, null, true)',
    )
  })

  test('expression with type', () => {
    const { code } = compileWithVOn(`<div onClick={handleClick as any} />`)
    expect(code).toMatchInlineSnapshot(`
      "
        const n0 = t0()
        n0.$evtclick = e => handleClick(e)
        return n0
      "
    `)
    expect(code).contains('n0.$evtclick = e => handleClick(e)')
  })
})
