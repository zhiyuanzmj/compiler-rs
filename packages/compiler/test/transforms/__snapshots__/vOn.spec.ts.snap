// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`v-on > event modifier 1`] = `
"
  const n1 = t0()
  const n3 = t1()
  const n5 = t0()
  const n7 = t2()
  const n9 = t2()
  const n11 = t0()
  const n13 = t2()
  const n15 = t3()
  const n17 = t3()
  const n19 = t3()
  const n21 = t3()
  const n23 = t3()
  const n25 = t3()
  const n27 = t3()
  const n29 = t3()
  const n31 = t3()
  const n33 = t3()
  const n35 = t3()
  const n37 = t3()
  const n39 = t3()
  const n41 = t3()
  const n43 = t3()
  n1.$evtclick = _withModifiers(e => handleEvent(e), ["stop"])
  _on(n3, "submit", _withModifiers(e => handleEvent(e), ["prevent"]))
  n5.$evtclick = _withModifiers(e => handleEvent(e), ["stop","prevent"])
  n7.$evtclick = _withModifiers(e => handleEvent(e), ["self"])
  _on(n9, "click", e => handleEvent(e), {
    capture: true
  })
  _on(n11, "click", e => handleEvent(e), {
    once: true
  })
  _on(n13, "scroll", e => handleEvent(e), {
    passive: true
  })
  n15.$evtcontextmenu = _withModifiers(e => handleEvent(e), ["right"])
  n17.$evtclick = _withModifiers(e => handleEvent(e), ["left"])
  n19.$evtmouseup = _withModifiers(e => handleEvent(e), ["middle"])
  n21.$evtcontextmenu = _withKeys(_withModifiers(e => handleEvent(e), ["right"]), ["enter"])
  n23.$evtkeyup = _withKeys(e => handleEvent(e), ["enter"])
  n25.$evtkeyup = _withKeys(e => handleEvent(e), ["tab"])
  n27.$evtkeyup = _withKeys(e => handleEvent(e), ["delete"])
  n29.$evtkeyup = _withKeys(e => handleEvent(e), ["esc"])
  n31.$evtkeyup = _withKeys(e => handleEvent(e), ["space"])
  n33.$evtkeyup = _withKeys(e => handleEvent(e), ["up"])
  n35.$evtkeyup = _withKeys(e => handleEvent(e), ["down"])
  n37.$evtkeyup = _withKeys(e => handleEvent(e), ["left"])
  n39.$evtkeyup = _withModifiers(e => submit(e), ["middle"])
  n41.$evtkeyup = _withModifiers(e => submit(e), ["middle","self"])
  n43.$evtkeyup = _withKeys(_withModifiers(e => handleEvent(e), ["self"]), ["enter"])
  return [n1, n3, n5, n7, n9, n11, n13, n15, n17, n19, n21, n23, n25, n27, n29, n31, n33, n35, n37, n39, n41, n43]
"
`;

exports[`v-on > namespace event with Component 1`] = `
"
  const n0 = _createComponent(Comp, { "onUpdate:modelValue": () => () => {} }, null, true)
  return n0
"
`;

exports[`v-on > should delegate event 1`] = `
"
  const n0 = t0()
  n0.$evtclick = e => test(e)
  return n0
"
`;

exports[`v-on > should not wrap keys guard if no key modifier is present 1`] = `
"
  const n0 = t0()
  n0.$evtkeyup = _withModifiers(e => test(e), ["exact"])
  return n0
"
`;

exports[`v-on > should support multiple modifiers and event options 1`] = `
"
  const n0 = t0()
  _on(n0, "click", _withModifiers(e => test(e), ["stop","prevent"]), {
    capture: true,
    once: true
  })
  return n0
"
`;

exports[`v-on > should transform click.middle 1`] = `
"
  const n0 = t0()
  n0.$evtmouseup = _withModifiers(e => test(e), ["middle"])
  return n0
"
`;

exports[`v-on > should transform click.right 1`] = `
"
  const n0 = t0()
  n0.$evtcontextmenu = _withModifiers(e => test(e), ["right"])
  return n0
"
`;

exports[`v-on > should use delegate helper when have multiple events of same name 1`] = `
"
  const n0 = t0()
  _delegate(n0, "click", e => test(e))
  _delegate(n0, "click", _withModifiers(e => test(e), ["stop"]))
  return n0
"
`;

exports[`v-on > should wrap keys guard for keyboard events or dynamic events 1`] = `
"
  const n0 = t0()
  _on(n0, "keydown", _withKeys(_withModifiers(e => test(e), ["stop","ctrl"]), ["a"]), {
    capture: true
  })
  return n0
"
`;

exports[`v-on > should wrap keys guard for static key event w/ left/right modifiers 1`] = `
"
  const n0 = t0()
  n0.$evtkeyup = _withKeys(e => test(e), ["left"])
  return n0
"
`;
